### PR TITLE
delete image from collection

### PR DESCRIPTION
**Pull request by tsop14**
Adds ability to remove an image from a collection from the image page. 
Updates UI in same way as removing labels - polls API until results match expectations, then broadcasts an event. 

remove-image-from-collection

### CODE REVIEW

File: kahuna/public/js/services/api/collections-api.js [hunk 1]
```diff
@@ -64,6 +66,41 @@ apiServices.factory('collections', ['mediaApi', function (mediaApi) {
...
+        return angular.equals(
+            collectionsA.sort(),
+            collectionsB.sort()
+        );
+    }
+
+    function getCollectionsIds(imageCollections) {
+        const collectionsIds = [];
+        imageCollections.data.forEach(e => collectionsIds.push(e.pathId));
+        return collectionsIds;
```
Comment (jamesgorrie) [inline kahuna/public/js/services/api/collections-api.js]: We could use a `map` here:
```JS
return imageCollections.data.map(col => col.pathId);
```

File: kahuna/public/js/services/image-accessor.js [hunk 2]
```diff
@@ -52,6 +52,18 @@ imageAccessor.factory('imageAccessor', function() {
         return userMetadata.data.archived.data;
     }
 
+    function readCollections(image) {
+        return image.data.collections;
+    }
+
+    function getCollectionsIds(image) {
```
Comment (jamesgorrie) [inline kahuna/public/js/services/image-accessor.js]: [See here](https://github.com/guardian/grid/pull/1620/files#r49436910)

File: kahuna/public/js/services/api/collections-api.js [hunk 3]
```diff
@@ -64,6 +66,41 @@ apiServices.factory('collections', ['mediaApi', function (mediaApi) {
...
+    }
+
+    function untilCollectionsEqual(image, expectedCollections) {
+        return image.get().then(apiImage => {
```
Comment (jamesgorrie) [inline kahuna/public/js/services/api/collections-api.js]: I know this seems to be the de facto way of doing this - but collections don't really need to poll as they don't affect image state - so I wonder if there is another way to update the collections.

Saying that, maybe it's worth leaving and doing in one refactor when we get to it @theefer?

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html [hunk 4]
```diff
@@ -202,6 +202,27 @@
     </div>
 </div>
 
+<div class="image-info" ng:if="ctrl.image.data.collections.length > 0">
+    <dl class="image-info__group ">
+        <dt class="metadata-line__key">Collections</dt>
+        <dd class="metadata-line__info flex-container"
+            ng-repeat="collection in ctrl.image.data.collections">
+            <a ui:sref="search.results({query: '~'+collection.data.pathId})">
+                {{collection.data.path.join(' ▸ ')}}
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html]: Are we sticking with this for now?

Comment (tsop14) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html]: Do you mean having it shown? Or the format of something?

File: kahuna/public/js/services/api/collections-api.js (hunk 3 above)
Comment (theefer) [inline kahuna/public/js/services/api/collections-api.js]: I'm guilty of having suggested that to @tsop14. It does affect the image state doesn't it, insofar as it requires updating the list of collections found on the image?

I'm going to look into the long-fabulated way to do this more properly without read-after-write and annoying polling, but I think this is the simplest way for now. If you have any other thoughts or suggestions, please share!

File: kahuna/public/js/services/api/collections-api.js [hunk 5]
```diff
@@ -64,6 +66,41 @@ apiServices.factory('collections', ['mediaApi', function (mediaApi) {
...
+                return apiImage;
+            } else {
+                return $q.reject();
+            }
+        });
+    }
+
+    function removeImageFromCollection(collection, image) {
+        return collection.perform('remove')
+            .then(newImageCollections => apiPoll(() =>
```
Comment (theefer) [inline kahuna/public/js/services/api/collections-api.js]: Does removing one collection return the list of all the remaining collections?

Comment (tsop14) [inline kahuna/public/js/services/api/collections-api.js]: Its returns the collections resource.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html (hunk 4 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html]: Just the arrow - I don't mind, I just can't remember.

File: kahuna/public/js/services/api/collections-api.js (hunk 5 above)
Comment (jamesgorrie) [inline kahuna/public/js/services/api/collections-api.js]: [It returns the new resource of collections](https://github.com/guardian/grid/blob/master/collections/app/controllers/ImageCollectionsController.scala#L54)

Comment (theefer) [inline kahuna/public/js/services/api/collections-api.js]: OK :+1:

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 6]
```diff
@@ -125,6 +127,12 @@ module.controller('grImageMetadataCtrl', [
                 });
         };
 
+        ctrl.removeImageFromCollection = function (collection) {
+            ctrl.removingCollection = collection;
+            collections.removeImageFromCollection(collection, ctrl.image)
+            .then(() => ctrl.removingCollection = false);
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js]: Indent for clarity?

**Comment by jamesgorrie** [comment]
:

**Comment by jamesgorrie** [comment]
:+1:

**Comment by tsop14** [comment]
@jamesgorrie would you mind checking this again? Thanks.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 7]
```diff
@@ -125,6 +136,12 @@ module.controller('grImageMetadataCtrl', [
                 });
         };
 
+        ctrl.removeImageFromCollection = function (collection) {
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js]: You could use `ctrl.removeImageFromCollection = collection =>  {`

**Comment by jamesgorrie** [comment]
:+1:

**Pull request merged**